### PR TITLE
Add unit-tests to `node::Taffy`

### LINKED ISSUE #177: Unit test `Node`
## What problem does this solve or what need does it fill?

`Node` should have unit-tests. ~~It does not currently have any.~~

Edit: Node has some tests in `tests/node.rs`. We need to verify that those tests are covering every behaviour we want it to.

## What solution would you like?

Add tests to `node` covering all its methods

### PR DESCRIPTION
**Pull request by Weibye**
# Objective

Fixes #177 

## Context

As I understand it:

- For a test to access private and internal behaviour, it needs to be in the same module as the definition it is trying to test.
- Unit-tests are to ensure correct and sound internal behavior.

So as a result I have moved all tests from `./tests/node.rs` to `./src/node.rs` and expanded upon them to cover more of the internals. I've created a discussion on this topic #179 where we can discuss this in detail.

## Changelog

- All tests relating to `node::Taffy` has been moved to `mod::node`
- `./test/node.rs` has been removed
- Added more tests to cover previously untested behavior

## Feedback wanted

- Overall discussion on where to keep what tests: #179 
- There are some tests that I currently have commented out since they don't really work but I'm not quite sure how to test for those, or if we need to test for those at all.

### CODE REVIEW

File: src/node.rs
Comment (Weibye) [inline src/node.rs:373]: These I don't know if are useful or not. They fail on `[Features:None]` as these are private methods

File: src/node.rs
Comment (TimJentzsch) [inline src/node.rs:15]: The idiomatic way to do tests in rust is to make a new module tests which is only compiled with `cfg(test)` and which contains all tests and test specific imports. That way, all the test functions are not compiled by default:

```rust
#[cfg(test)]
mod tests {
    use crate::style::{Dimension, Display, FlexDirection};
    
    // Test functions
}
```

File: src/node.rs
Comment (TimJentzsch) [inline src/node.rs:536]: I _think_ it's more idiomatic to have the comment above the `#[test]`, but I'm not 100% sure on that

**Review by TimJentzsch — CHANGES_REQUESTED** [review]
Please move the tests to a `tests` module (in the same file). You can take a look at the `MaybeMath` tests as reference.

The tests itself look good to me though.

File: src/node.rs
Comment (Weibye) [reply src/node.rs:15]: As outlined in https://github.com/DioxusLabs/taffy/discussions/179, wouldn't that make it impossible to test internal and private behavior?

Comment (Weibye) [reply src/node.rs:15]: I agree that moving them to `mod tests` is cleaner, and this is why I'm struggling a bit to understand exactly what is the best place for the tests to be (hence the discussion)

Comment (TimJentzsch) [reply src/node.rs:15]: Because `mod tests` is inside the same module, it should have access to the private functions.

Does [this post](https://users.rust-lang.org/t/testing-private-functions/41097/3) help?

Comment (Weibye) [reply src/node.rs:15]: Ah, in that case that is much better 💯 Thanks! This was the missing piece :)

Comment (Weibye) [reply src/node.rs]: Done :)

File: src/node.rs
Comment (Weibye) [reply src/node.rs]: Done

File: src/node.rs
Comment (TimJentzsch) [inline src/node.rs:342]: I think we don't need this one

File: src/node.rs
Comment (TimJentzsch) [inline src/node.rs:384]: Do you know why `contains_key` is private with [Features: None]? That looks like a bug to me.

File: src/node.rs
Comment (TimJentzsch) [inline src/node.rs:398]: This is not for this PR, but I feel like we should add a function that creates a new node without children, since it's a pretty common pattern.

File: src/node.rs
Comment (TimJentzsch) [inline src/node.rs:656]: This will need to be changed to be a `sys::Vec` instead of `std::vec::Vec`

File: src/node.rs
Comment (Weibye) [reply src/node.rs:398]: Yeah I had the same thought as I was creating this 👍

File: src/node.rs
Comment (Weibye) [reply src/node.rs]: Removed

File: src/node.rs
Comment (Weibye) [reply src/node.rs:381]: Created #180

File: src/node.rs
Comment (Weibye) [reply src/node.rs:370]: I have no idea :/ Maybe @alice-i-cecile have some insight?

Comment (Weibye) [reply src/node.rs:370]: Ah, hang on, `IndexMap.contains_key()` needs to be `pub` for this to work I think.

Comment (Weibye) [reply src/node.rs:370]: This was solved by exposing the methods on `IndexMap`

File: src/node.rs
Comment (Weibye) [reply src/node.rs]: Fixed by manually pushing to the vector instead of using the macro to create it.

File: src/node.rs
Comment (TimJentzsch) [inline src/node.rs:728]: Is this still needed?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Happy to merge once Tim's comment is resolved.

File: src/node.rs
Comment (Weibye) [reply src/node.rs:728]: I'll remove it. The only path we are not covering atm is that anything that can produce an error is actually doing so. 

I think we can add those later once we have more tests overall and better tooling for seeing what coverage looks like.

**Comment by TimJentzsch** [comment]
As a tip: You can use `cargo test --no-default-features` locally to run the tests without default features enabled, then you don't have to wait for CI every time (and it might clean up the commits a bit) :)

**Comment by Weibye** [comment]
> As a tip: You can use `cargo test --no-default-features` locally to run the tests without default features enabled, then you don't have to wait for CI every time (and it might clean up the commits a bit) :)

That I didn't know, thanks!

**Review by Weibye — COMMENTED** [review]
Hang on, going to migrate tests to new_leaf

**Review by alice-i-cecile — APPROVED** [review]

**Comment by Weibye** [comment]
All comments should now be resolved :)

**Review by TimJentzsch — APPROVED** [review]

**Pull request merged**
